bee-tangle/src/helper.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright 2020 IOTA Stiftung
```

Comment:
I would put that either in `bee-test` or in a `tests` folder in the `bee-tangle` crate. This is totally tests related, it's a bit misleading to  call it `helper` or even to put it in the `src` folder IMO.

Reply:
let me explain:
`bee-test` is targeted to provide test utilities that may be useful for many different crates, like e.g. providing relatively realistic transactions. This is not the case here. This module contains specific test helper functions of this crate.

The `tests` folder is **only** for integration tests which serve a completely different purpose. Unit tests reside in `src`. Furthermore, under the `src` directory are many test modules already. They just don't have their own file. So what I did was just to go a different route here (but it's totally not unconventional or misleading). I just need to have a trivial tangle in several of my modules. I could leave it one and let the other link it in. But that would be really misleading as there is no reason for a preference.

Reply:
we could call it differently though :smile:

Reply:
> bee-test is targeted at providing test utilities that may be useful for many different crates, like e.g. providing relatively realistic transactions. This is not the case here. This module contains specific test helper functions shared only in this crate.

We made the tangle generic so that it can be used/implemented in different crates. Providing realistic transactions and realistic tangles is to me the exact same thing.

> The tests folder is only for integration tests which serve an orthogonal purpose.

If this is an absolute rule, then the ternary crate doesn't respect it, I see a lot of unit tests in its `tests` folder.

> If it's that disturbing to you

Yes it is disturbing and misleading to. I definitely expect to find tangle helpers in this module.

Reply:
> We made the tangle generic so that it can be used/implemented in different crates.

sure, but the more specific tangles like the milestone-based tangle need different setups, so they most certainly don't need the **same** test helpers as the most generic tangle. Just look at white-flag tangle set up you created. Currently only one crate `bee-protocol` builds on top of the generic tangle. Things can be moved easily if there is the need for.

> If this is an absolute rule, then the ternary crate doesn't respect it, I see a lot of unit tests in its tests folder.

If you are sure those are unit tests then this are misplaced indeed. 
[the book](https://doc.rust-lang.org/stable/book/ch11-03-test-organization.html?highlight=integration#integration-tests)
[rust by example](https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html)

Reply:
But since those compile, they obviously only test the public API, hence are integration tests. Unit tests need to have access to the private functions of a module.

Reply:
So why are tangle traversal tests not integration tests ?

Reply:
That's a good question actually :smile: 

I am not sure about the consequences this train of thoughts has as it would mean that all public methods/functions should only be tested as an integration test.

Reply:
In general if the test only checks a single function I use a unit test for it, whether it's private or not. For testing overall functionality/behavior (not interested in the details) I would tend to use an integration test.

Reply:
> it would mean that all public methods/functions should only be tested as an integration test.

Doesn't sound so bad to me tbh :)

> For testing overall functionality/behavior (not interested in the details) I would tend to use an integration test.

I believe we would agree than tangle traversals are an overall functionality compared to individual features of the tangle.

Anyway let's continue tomorrow!

Reply:
I think I agree with you on this. because a good framework has its API well documented, not only with plain text, but also at least in Rust with `doc tests`, and guess what doc tests essentially are. ;)

Reply:
Yeah nice discussion, I probably need to review my own approach to tests :)

Reply:
(because I also have a lot of integration tests in src modules)